Lib/test/test_capi/test_number.py

```diff
@@ -0,0 +1,432 @@
+import unittest
+import sys
+import warnings
+
+from test.support import import_helper
+from test.test_capi.test_getargs import (Float, BadFloat, BadFloat2, Index,
+                                         BadIndex, BadIndex2, Int, BadInt, BadInt2)
+
+_testcapi = import_helper.import_module('_testcapi')
+from _testcapi import PY_SSIZE_T_MIN, PY_SSIZE_T_MAX
+
+NULL = None
+
+
+class BadAdd:
+    def __add__(self, other):
+        raise RuntimeError
+
+
+class BadAdd2(int):
+    def __radd__(self, other):
+        return NotImplemented
+
+
+class IndexLike:
+    def __init__(self, value):
+        self.value = value
+
+    def __index__(self):
+        return self.value
+
+
+class CAPITest(unittest.TestCase):
+    def test_check(self):
+        # Test PyNumber_Check()
+        check = _testcapi.number_check
+
+        self.assertTrue(check(1))
+        self.assertTrue(check(1.0))
+        self.assertTrue(check(1+2j))
+        self.assertFalse(check([]))
+        self.assertFalse(check(object()))
+        self.assertFalse(check(NULL))
+
+    def test_add(self):
+        # Test PyNumber_Add()
+        add = _testcapi.number_add
+
+        self.assertEqual(add(1, 2), 1 + 2)
+        self.assertEqual(add(1, 0.75), 1 + 0.75)
+        self.assertEqual(add(0.75, 1), 0.75 + 1)
+        self.assertEqual(add(1, BadAdd2(2)), 3)
+        self.assertEqual(add([1, 2], [3, 4]), [1, 2, 3, 4])
+
+        self.assertRaises(TypeError, add, 1, object())
+        self.assertRaises(TypeError, add, object(), 1)
+
+        # CRASHES add(NULL, 42)
+        # CRASHES add(42, NULL)
+
+    def test_subtract(self):
+        # Test PyNumber_Subtract()
+        subtract = _testcapi.number_subtract
+
+        self.assertEqual(subtract(1, 2), 1 - 2)
+
+        self.assertRaises(TypeError, subtract, 1, object())
+        self.assertRaises(TypeError, subtract, object(), 1)
+
+        # CRASHES subtract(NULL, 42)
+        # CRASHES subtract(42, NULL)
+
+    def test_multiply(self):
+        # Test PyNumber_Multiply()
+        multiply = _testcapi.number_multiply
+
+        self.assertEqual(multiply(2, 3), 2 * 3)
+        self.assertEqual(multiply([1], 2), [1, 1])
+        self.assertEqual(multiply(2, [1]), [1, 1])
+
+        self.assertRaises(TypeError, multiply, 1, object())
+        self.assertRaises(TypeError, multiply, object(), 1)
+
+        # CRASHES multiply(NULL, 42)
+        # CRASHES multiply(42, NULL)
+
+    def test_matrixmultiply(self):
```

Comment:
Look how `__matmul__` is tested in other tests. The all define a test class with `__matmul__`, `__rmatmul__`, `__imatmul__`.

Reply:
Yeah, I did something like.